File: scripts/interpolator_controller.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
import numpy as np
import rclpy.time

from ros_interface_mpc.msg import Trajectory, InitialState
from rclpy.qos import QoSProfile
from rclpy.time import Time
from ros_interface_mpc_utils.conversions import multiarray_to_numpy


from nav_msgs.msg import Odometry
from std_msgs.msg import Float64MultiArray, Float64

from go2_control_interface.robot_interface import Go2RobotInterface
from proxsuite_nlp import manifolds
from simple_mpc import IDSolver, RobotHandler
import example_robot_data

class InterpolatorControllerNode(Node):

    def __init__(self):
        # Initialization of node
        super().__init__('interpolator_controller')
        self.declare_parameter('mpc_type')
        self.parameter = self.get_parameter('mpc_type')
        self.robotIf = Go2RobotInterface(self)
        self.start_mpc = False

        # Load robot
        self.rmodel = example_robot_data.load("go2").model
        SRDF_SUBPATH = "/go2_description/srdf/go2.srdf"
        URDF_SUBPATH = "/go2_description/urdf/go2.urdf"
        modelPath = example_robot_data.getModelPath(URDF_SUBPATH)

        design_conf = dict(
            urdf_path=modelPath + URDF_SUBPATH,
            srdf_path=modelPath + SRDF_SUBPATH,
            robot_description="",
            root_name="root_joint",
            base_configuration="standing",
            controlled_joints_names=[
                "root_joint",
                "FL_hip_joint",
                "FL_thigh_joint",
                "FL_calf_joint",
                "FR_hip_joint",
                "FR_thigh_joint",
                "FR_calf_joint",
                "RL_hip_joint",
                "RL_thigh_joint",
                "RL_calf_joint",
                "RR_hip_joint",
                "RR_thigh_joint",
                "RR_calf_joint",
            ],
            end_effector_names=[
                "FL_foot",
                "FR_foot",
                "RL_foot",
                "RR_foot",
            ],
            hip_names=[
                "FL_thigh",
                "FR_thigh",
                "RL_thigh",
                "RR_thigh",
            ],
            feet_to_base_trans=[
                np.array([0.2, 0.1, 0.]),
                np.array([0.2, -0.1, 0.]),
                np.array([-0.2, 0.1, 0.]),
                np.array([-0.2, -0.1, 0.]),
            ]
        )
        self.handler = RobotHandler()
        self.handler.initialize(design_conf)

        # Define state publisher
        qos_profile_keeplast = QoSProfile(history=rclpy.qos.HistoryPolicy.KEEP_LAST, depth=1)
        self.robot_pub = self.create_publisher(InitialState, 'initial_state', qos_profile_keeplast)

        # Define command subscriber
        self.subscription_joints = self.create_subscription(
            Trajectory,
            'trajectory',
            self.trajectory_callback,
            qos_profile_keeplast)

        # Define odometry subscriber
        self.subscription_odom = self.create_subscription(
            Odometry,
            'odometry/filtered',
            self.odometry_callback,
            qos_profile_keeplast)

        # Message declarations for torque
        self.torque_simu = np.zeros(18)

        self.space = manifolds.MultibodyPhaseSpace(self.rmodel)
        self.ndx = 36
        self.nu = 12
        self.nq = 19
        self.nv = 18

        # Define a default configuration
        self.default_standing_q = np.array([
            0., 0., 0.335, 0., 0., 0., 1.,
            0.0899, 0.8130, -1.596,
            -0.0405, 0.824, -1.595,
            0.1695, 0.824, -1.606,
            -0.1354, 0.820, -1.593
        ])
        self.default_standing_v = np.zeros(self.nv)
        self.default_standing_u = np.array([
            -3.71, -1.81,  5.25,
            3.14, -1.37, 5.54,
            -1.39, -1.09,  3.36,
            1.95, -0.61,  3.61
        ])

        # Start state
        self.trajectoryStamp = self.get_clock().now()
        self.x0 = np.concatenate((self.default_standing_q, self.default_standing_v))
        self.u0 = self.default_standing_u.copy()
        self.K0 = np.zeros((self.nu, self.ndx))

        self.Kp = [150.]*self.nu
        self.Kd = 0.25 * np.sqrt(self.Kp)
        self.jointCommand = self.default_standing_q[7:]
        self.velocityCommand = self.default_standing_v[6:]
        self.torqueCommand = self.default_standing_u.copy()
        self.MPC_timestep = 0.01

        # For filtering base position and velocity
        self.t_odom_update = None
        self.base_pose = [0.,0.,0., 0.,0.,0.,1.]
        self.base_vel = [0.,0.,0., 0.,0.,0.]
        self.base_filter_fq = self.declare_parameter("base_filter_fq", -1.0).value # By default no filter

        # Debugging purposes
        self.debug_filter_pub = self.create_publisher(Float64MultiArray, "/debug/filtered_state", 10)
        self.debug_loop_time_pub = self.create_publisher(Float64, "/debug/loop_time", 10)
        self.debug_loop_time_msg = Float64()
        self.debug_comm_time_pub = self.create_publisher(Float64, "/debug/comm_time", 10)
        self.debug_comm_time_msg = Float64()
        self.debug_comm_time_new_msg = False

        """ Initialize whole-body inverse dynamics QP"""
        id_conf = dict(
            contact_ids=self.handler.getFeetIds(),
            x0=self.x0,
            mu=0.8,
            Lfoot=0.01,
            Wfoot=0.01,
            force_size=3,
            kd=0,
            w_force=100,
            w_acc=1,
            w_tau=0,
            verbose=False,
        )

        self.qp = IDSolver(id_conf, self.rmodel)

        self.robotIf.start_async(self.default_standing_q[7:].tolist())
        self.robotIf.register_callback(self.control_loop)

    def trajectory_callback(self, msg):
        self.us = multiarray_to_numpy(msg.us)
        self.xs = multiarray_to_numpy(msg.xs)
        self.K0 = multiarray_to_numpy(msg.k0)
        self.trajectoryStamp = Time.from_msg(msg.stamp)
        self.ddqs = multiarray_to_numpy(msg.ddqs)
        self.contact_states = multiarray_to_numpy(msg.contact_states)
        self.forces = multiarray_to_numpy(msg.forces)

        self.start_mpc = True
        self.debug_comm_time_new_msg = True
        self.debug_comm_time_process_time = msg.process_duration

    def odometry_callback(self, msg):
        t_meas = rclpy.time.Time.from_msg(msg.header.stamp).nanoseconds * 1e-9

        pos_meas = msg.pose.pose.position.x, \
                   msg.pose.pose.position.y, \
                   msg.pose.pose.position.z
        quat_meas = msg.pose.pose.orientation.x, \
                    msg.pose.pose.orientation.y, \
                    msg.pose.pose.orientation.z, \
                    msg.pose.pose.orientation.w

        v_meas = msg.twist.twist.linear.x, \
                 msg.twist.twist.linear.y, \
                 msg.twist.twist.linear.z, \
                 msg.twist.twist.angular.x, \
                 msg.twist.twist.angular.y, \
                 msg.twist.twist.angular.z

        # Filter base position and orientation
        b = 0.
        if self.t_odom_update is not None and self.base_filter_fq > 0.:
            b = 1. / (1 + 2 * 3.14 * (t_meas - self.t_odom_update) * self.base_filter_fq)

        self.base_pose[:3] = [(1-b) * pos_meas[i] + b * self.base_pose[i] for i in range(3)]
        self.base_pose[3:] = [quat_meas[i] for i in range(4)]
        self.base_vel      = [(1-b) * v_meas[i]   + b * self.base_vel[i] for i in range(6)]
        self.t_odom_update = t_meas

    # Main control loop
    def control_loop(self,t, q, v, a):
        delay = t - self.trajectoryStamp.nanoseconds * 1e-9
        x_measured = np.concatenate((self.base_pose, q, self.base_vel, v))

        if self.start_mpc:
            self.Kp = [10.]*self.nu
            self.Kd = (np.sqrt(self.Kp)).tolist()

            step_nb = int(delay // self.MPC_timestep)
            step_progress = (delay % self.MPC_timestep) / self.MPC_timestep
            if(step_nb >= len(self.xs) -1):
                step_nb = len(self.xs) - 1
                step_progress = 0.0
                x_interpolated = self.xs[step_nb]
                u_interpolated = self.us[step_nb]
            else:
                x_interpolated = self.xs[step_nb + 1] * step_progress  + self.xs[step_nb] * (1. - step_progress)
                u_interpolated = self.us[step_nb + 1] * step_progress  + self.us[step_nb] * (1. - step_progress)

            self.jointCommand = x_interpolated[7:self.nq]
            self.velocityCommand = x_interpolated[self.nq + 6:]
            if self.parameter.value == "fulldynamics":
                self.torqueCommand = u_interpolated - 1.0 * self.K0 @ self.space.difference(x_measured, x_interpolated)

            elif self.parameter.value == "kinodynamics":
                self.Kp = [50.]*self.nu
                self.Kd = (np.sqrt(self.Kp)).tolist()
                if(step_nb >= len(self.xs) -1):
                    a_interpolated = self.ddqs[step_nb]
                    forces_interpolated = self.forces[step_nb]
                else:
                    a_interpolated = self.ddqs[step_nb + 1] * step_progress  + self.ddqs[step_nb] * (1. - step_progress)
                    forces_interpolated = self.forces[step_nb + 1] * step_progress  + self.forces[step_nb] * (1. - step_progress)
                self.handler.updateState(x_measured[:self.nq], x_measured[self.nq:], True)
                self.qp.solve_qp(
                    self.handler.getData(),
                    [bool(self.contact_states[step_nb][i]) for i in range(4)],
                    x_measured[self.nq:],
                    a_interpolated,
                    forces_interpolated,
                    self.handler.getMassMatrix(),
                )
                self.torqueCommand = self.qp.solved_torque


        if (self.robotIf.is_ready):
            self.robotIf.send_command(self.jointCommand.tolist(),
                                    self.velocityCommand.tolist(),
                                    self.torqueCommand.tolist(),
                                    self.Kp, #self.Kp.tolist(),
                                    self.Kd #self.Kd.tolist()
            )

            state = InitialState()
            state.stamp = rclpy.time.Time(seconds=t).to_msg()
            state.x0 = x_measured.tolist()
            self.robot_pub.publish(state)

        ############################
        # Un-comment for debugging #
        ############################

        # # Log delay
        # self.debug_loop_time_msg.data = delay
        # self.debug_loop_time_pub.publish(self.debug_loop_time_msg)

        # # Log filtered state
        # debug_msg = Float64MultiArray()
        # debug_msg.data = x_measured.ravel().tolist()
        # self.debug_filter_pub.publish(debug_msg)

        # Log communication time
        # if(self.debug_comm_time_new_msg):
        #     self.debug_comm_time_new_msg = False
        #     self.debug_comm_time_msg.data = delay - self.debug_comm_time_process_time
        #     self.debug_comm_time_pub.publish(self.debug_comm_time_msg)

def main(args=None):
    rclpy.init(args=args)

    mpc_subscriber = InterpolatorControllerNode()
    rclpy.spin(mpc_subscriber)

    mpc_subscriber.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/interpolator_controller.cpp
// Copyright 2016 Open Source Robotics Foundation, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <functional>
#include <memory>

#include <rclcpp/rclcpp.hpp>
#include <rmw/qos_profiles.h>
#include <rclcpp/qos.hpp>
#include <nav_msgs/msg/odometry.hpp>   // for the tf transformation
#include <Eigen/Core>
#include <Eigen/Geometry>
#include <proxsuite-nlp/modelling/spaces/multibody.hpp>
#include <simple-mpc/robot-handler.hpp>
#include <simple-mpc/lowlevel-control.hpp>
#include <pinocchio/fwd.hpp>
#include <pinocchio/parsers/srdf.hpp>
#include <pinocchio/parsers/urdf.hpp>
#include <pinocchio/algorithm/model.hpp>
#include <pinocchio/multibody/model.hpp>

#include "ros_interface_mpc/msg/trajectory.hpp"
#include "ros_interface_mpc/msg/initial_state.hpp"

using std::placeholders::_1;

class InterpolatorController : public rclcpp::Node
{
public:
  InterpolatorController()
  : Node("interpolator_controller")
  {
    this->declare_parameter("mpc_type", rclcpp::PARAMETER_STRING);
    this->declare_parameter("base_filter_fq", rclcpp::PARAMETER_DOUBLE);
    mpc_type_ = this->get_parameter("mpc_type").as_string();
    base_filter_fq_ = this->get_parameter("base_filter_fq").as_double();
    start_mpc_ = false;

    // Load robot
    pinocchio::Model model;
    std::string urdf_path = EXAMPLE_ROBOT_DATA_MODEL_DIR "/go2_description/urdf/go2.urdf";
    std::string srdf_path = EXAMPLE_ROBOT_DATA_MODEL_DIR "/go2_description/srdf/go2.srdf";
    std::string base_joint_name ="root_joint";
    
    pinocchio::urdf::buildModel(urdf_path, pinocchio::JointModelFreeFlyer(), model);
    pinocchio::srdf::loadReferenceConfigurations(model, srdf_path, false);
    //pinocchio::srdf::loadRotorParameters(model, srdf_path, false);

    simple_mpc::RobotModelHandler model_handler = 
      simple_mpc::RobotModelHandler(model, "standing", base_joint_name);
    
    /// Add reference foot for walking
    pinocchio::SE3 ref_FL_foot = pinocchio::SE3::Identity();
    pinocchio::SE3 ref_FR_foot = pinocchio::SE3::Identity();
    pinocchio::SE3 ref_RL_foot = pinocchio::SE3::Identity();
    pinocchio::SE3 ref_RR_foot = pinocchio::SE3::Identity();
    ref_FL_foot.translation() = Eigen::Vector3d(0.17, 0.15, 0.0);
    ref_FR_foot.translation() = Eigen::Vector3d(0.17, -0.15, 0.0);
    ref_RL_foot.translation() = Eigen::Vector3d(-0.24, 0.15, 0.0); 
    ref_RR_foot.translation() = Eigen::Vector3d(-0.24, -0.15, 0.0);
    model_handler.addFoot("FL_foot", base_joint_name, ref_FL_foot);
    model_handler.addFoot("FR_foot", base_joint_name, ref_FR_foot);
    model_handler.addFoot("RL_foot", base_joint_name, ref_RL_foot);
    model_handler.addFoot("RR_foot", base_joint_name, ref_RR_foot);
    
    // Initialize model related objects
    data_handler_ = std::make_shared<simple_mpc::RobotDataHandler>(model_handler);
    space_ = std::make_shared<proxsuite::nlp::MultibodyPhaseSpace<double>>(model_handler.getModel());
    nq_ = model_handler.getModel().nq;
    nv_ = model_handler.getModel().nv;
    nu_ = nv_ - 6;

    // Initialize vectors and matrices
    t_odom_update_ = 0;
    base_pose_.resize(7);
    base_pose_ << 0, 0, 0, 0, 0, 0, 1; 
    base_vel_.resize(6);
    base_vel_ << 0, 0, 0, 0, 0, 0; 
    
    // Default standing base and joints position for Go2
    Eigen::VectorXd default_standing_q(nq_);
    default_standing_q << 0., 0., 0.335, 0., 0., 0., 1.,
      0.0899, 0.8130, -1.596,
      -0.0405, 0.824, -1.595,
      0.1695, 0.824, -1.606,
      -0.1354, 0.820, -1.593;

    joint_command_.resize(nu_);
    joint_command_ = default_standing_q.tail(nu_);
    velocity_command_.resize(nu_);
    velocity_command_.setZero();
    torque_command_.resize(nu_);
    torque_command_ << -3.71, -1.81, 5.25,
      3.14, -1.37, 5.54,
      -1.39, -1.09,  3.36,
      1.95, -0.61,  3.61;
    x_measured_.resize(nq_ + nv_);
    x_measured_ << default_standing_q, Eigen::VectorXd::Zero(nv_);
    x_interpolated_.resize(nq_ + nv_);
    u_interpolated_.resize(nu_);
    a_interpolated_.resize(nv_);
    forces_interpolated_.resize(model_handler.getFeetNames().size() * 3);
    
    Kp_ = Eigen::VectorXd::Constant(nu_, 150);
    Kd_ = Eigen::VectorXd::Constant(nu_, 0.5 * sqrt(Kp_[0])); 
    MPC_timestep_ = 0.01;

    // Define state publisher
    trajectory_pub_ = this->create_publisher<ros_interface_mpc::msg::InitialState>(
      "initial_state", rclcpp::QoS(rclcpp::KeepLast(1), rmw_qos_profile_sensor_data));

    // Define subscriber
    subscription_joints_ = this->create_subscription<ros_interface_mpc::msg::Trajectory>(
      "trajectory", 
      rclcpp::QoS(rclcpp::KeepLast(1), rmw_qos_profile_sensor_data), 
      std::bind(&InterpolatorController::trajectory_callback, this, std::placeholders::_1)
    );

    subscription_odom_ = this->create_subscription<nav_msgs::msg::Odometry>(
      "odometry/filtered", 
      rclcpp::QoS(rclcpp::KeepLast(1), rmw_qos_profile_sensor_data), 
      std::bind(&InterpolatorController::odometry_callback, this, std::placeholders::_1)
    );

    // Define inverse dynamics QP
    simple_mpc::IDSettings settings;
    settings.contact_ids=model_handler.getFeetIds();
    settings.mu=0.8;
    settings.Lfoot=0.01;
    settings.Wfoot=0.01;
    settings.force_size=3;
    settings.kd=0;
    settings.w_force=100;
    settings.w_acc=1;
    settings.w_tau=0;
    settings.verbose=false;

    qp_ = std::make_shared<simple_mpc::IDSolver>(settings, model_handler.getModel());
  }

private:
  std::vector<Eigen::VectorXd> floatMultiArrayToVector(
    const std_msgs::msg::Float64MultiArray& multiArray)
  {
    std::vector<Eigen::VectorXd> data;
    data.resize(multiArray.layout.dim[0].size, Eigen::VectorXd::Zero(multiArray.layout.dim[1].size)); 
    unsigned int vec_id = 0;
    unsigned int eigen_id = 0;
    for (size_t i = 0; i < multiArray.data.size(); ++i) {
      data[vec_id][eigen_id] = multiArray.data[i];
      eigen_id ++;
      if (eigen_id == multiArray.layout.dim[1].size) {
        eigen_id = 0;
        vec_id ++;
      }
    }
    return data;
  }

  std::vector<std::vector<bool>> intMultiArrayToVector(
    const std_msgs::msg::Int8MultiArray& multiArray)
  {
    std::vector<std::vector<bool>> data;
    data.resize(multiArray.layout.dim[0].size, std::vector<bool>(multiArray.layout.dim[1].size)); 
    unsigned int vec_id = 0;
    unsigned int eigen_id = 0;
    for (size_t i = 0; i < multiArray.data.size(); ++i) {
      data[vec_id][eigen_id] = static_cast<bool>(multiArray.data[i]);
      eigen_id ++;
      if (eigen_id == multiArray.layout.dim[1].size) {
        eigen_id = 0;
        vec_id ++;
      }
    }
    return data;
  }

  Eigen::MatrixXd floatMultiArrayToMatrix(
    const std_msgs::msg::Float64MultiArray& multiArray)
  {
    Eigen::MatrixXd data = 
      Eigen::MatrixXd::Zero(multiArray.layout.dim[0].size, multiArray.layout.dim[1].size);
    unsigned int line = 0;
    unsigned int col = 0;
    for (size_t i = 0; i < multiArray.data.size(); ++i) {
      data(line, col) = multiArray.data[i];
      col ++;
      if (col == multiArray.layout.dim[1].size) {
        col = 0;
        line ++;
      }
    }
    return data;
  } 
  
  void trajectory_callback(const ros_interface_mpc::msg::Trajectory::SharedPtr msg)
  {
    start_mpc_ = true;
    us_ = floatMultiArrayToVector(msg->us);
    xs_ = floatMultiArrayToVector(msg->xs);
    K0_ = floatMultiArrayToMatrix(msg->k0);
    trajectoryStamp_ = rclcpp::Time(msg->stamp.sec, msg->stamp.nanosec);
    ddqs_ = floatMultiArrayToVector(msg->ddqs);
    contact_states_ = intMultiArrayToVector(msg->contact_states);
    forces_ = floatMultiArrayToVector(msg->forces); 
  }

  void odometry_callback(const nav_msgs::msg::Odometry::SharedPtr msg)
  {
    auto t_meas = msg->header.stamp.sec + msg->header.stamp.nanosec * 1e-9;

    // Filter base position and orientation
    double b = 0.;
    if (t_odom_update_ > 0 and base_filter_fq_ > 0.) {
      b = 1. / (1 + 2 * 3.14 * (t_meas - t_odom_update_) * base_filter_fq_);
    }

    base_pose_[0] = (1-b) * msg->pose.pose.position.x + b * base_pose_[0];
    base_pose_[1] = (1-b) * msg->pose.pose.position.y + b * base_pose_[1];
    base_pose_[2] = (1-b) * msg->pose.pose.position.z + b * base_pose_[2];

    base_pose_[3] = msg->pose.pose.orientation.x;
    base_pose_[4] = msg->pose.pose.orientation.y;
    base_pose_[5] = msg->pose.pose.orientation.z;
    base_pose_[6] = msg->pose.pose.orientation.w;

    base_vel_[0] = (1-b) * msg->twist.twist.linear.x + b * base_vel_[0];
    base_vel_[1] = (1-b) * msg->twist.twist.linear.y + b * base_vel_[1];
    base_vel_[2] = (1-b) * msg->twist.twist.linear.z + b * base_vel_[2];
    base_vel_[3] = (1-b) * msg->twist.twist.angular.x + b * base_vel_[3];
    base_vel_[4] = (1-b) * msg->twist.twist.angular.y + b * base_vel_[4];
    base_vel_[5] = (1-b) * msg->twist.twist.angular.z + b * base_vel_[5];

    t_odom_update_ = t_meas; 
  }
  
  void control_loop(double& t, const Eigen::VectorXd& q, const Eigen::VectorXd& v, const Eigen::VectorXd& a) {
    // Compute control delay
    double delay = t - trajectoryStamp_.nanoseconds() * 1e-9;

    // Concatenate odometry and joint measures
    x_measured_ << base_pose_, q, base_vel_, v;
    
    if (start_mpc_) {
      for (size_t i = 0; i < nu_; i++) {
        Kp_[i] = 100.;
        Kd_[i] = 0.5 * sqrt(Kp_[i]);
      }
      
      // Compute the time knot corresponding to the current delay
      int step_nb = static_cast<int>(delay / MPC_timestep_);
      double step_progress = (delay - step_nb * MPC_timestep_) / MPC_timestep_;
      
      // Interpolate state and command trajectories
      if (step_nb >= (int)xs_.size() -1) {
        step_nb = xs_.size() - 1;
        step_progress = 0.0;
        x_interpolated_ = xs_[step_nb];
        u_interpolated_ = us_[step_nb];
      }
      else {
        x_interpolated_ = xs_[step_nb + 1] * step_progress  + xs_[step_nb] * (1. - step_progress);
        u_interpolated_ = us_[step_nb + 1] * step_progress  + us_[step_nb] * (1. - step_progress);
      }

      joint_command_ = x_interpolated_.segment(7, nu_);
      velocity_command_ = x_interpolated_.tail(nu_);

      // Compute torque command depending on MPC type
      if (mpc_type_ == "fulldynamics") 
        torque_command_ = u_interpolated_ - 1.0 * K0_ * space_->difference(x_measured_, x_interpolated_);
      else if (mpc_type_ == "kinodynamics") {
        for (size_t i = 0; i < nu_; i++) {
          Kp_[i] = 50.;
          Kd_[i] = 0.5 * sqrt(Kp_[i]);
        }

        // Interpolate acceleration and force trajectories
        if (step_nb >= (int)xs_.size() -1) {
          a_interpolated_ = ddqs_[step_nb];
          forces_interpolated_ = forces_[step_nb];
        }
        else {
          a_interpolated_ = ddqs_[step_nb + 1] * step_progress  + ddqs_[step_nb] * (1. - step_progress);
          forces_interpolated_ = forces_[step_nb + 1] * step_progress  + forces_[step_nb] * (1. - step_progress);
        }

        data_handler_->updateInternalData(x_measured_, true);
        pinocchio::Data data = data_handler_->getData();

        // Solve inverse dynamics QP
        qp_->solveQP(
          data,
          contact_states_[step_nb],
          x_measured_.tail(nv_),
          a_interpolated_,
          Eigen::VectorXd::Zero(nu_),
          forces_interpolated_,
          data_handler_->getData().M
        );
        torque_command_ = qp_->solved_torque_;
      }
    }
  }
  
  double t_odom_update_;
  double base_filter_fq_;
  double MPC_timestep_;
  size_t nu_;
  size_t nv_;
  size_t nq_;
  std::string mpc_type_;
  bool start_mpc_;
  rclcpp::Time trajectoryStamp_;

  // Trajectory vectors updated by the OCP solver
  std::vector<Eigen::VectorXd> us_;
  std::vector<Eigen::VectorXd> xs_;
  std::vector<Eigen::VectorXd> ddqs_;
  std::vector<Eigen::VectorXd> forces_;
  std::vector<std::vector<bool>> contact_states_;
  Eigen::MatrixXd K0_;

  // Model-related quantities
  std::shared_ptr<proxsuite::nlp::MultibodyPhaseSpace<double>> space_;
  std::shared_ptr<simple_mpc::RobotDataHandler> data_handler_;
  std::shared_ptr<simple_mpc::IDSolver> qp_;
  
  // Intermediate vectors
  Eigen::VectorXd base_pose_;
  Eigen::VectorXd base_vel_;
  Eigen::VectorXd x_measured_;
  Eigen::VectorXd x_interpolated_;
  Eigen::VectorXd u_interpolated_;
  Eigen::VectorXd a_interpolated_;
  Eigen::VectorXd forces_interpolated_;

  // Command vectors
  Eigen::VectorXd joint_command_;
  Eigen::VectorXd velocity_command_;
  Eigen::VectorXd torque_command_;
  Eigen::VectorXd Kp_;
  Eigen::VectorXd Kd_;

  // ROS topics
  rclcpp::Subscription<ros_interface_mpc::msg::Trajectory>::SharedPtr subscription_joints_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr subscription_odom_;
  rclcpp::Publisher<ros_interface_mpc::msg::InitialState>::SharedPtr trajectory_pub_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<InterpolatorController>());
  rclcpp::shutdown();
  return 0;
}
